File: typescript/src/giftcard/application/services/bookingResultCrediting.ts
// Servizio applicativo che reagisce ai risultati della prenotazione e accredita la gift card.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Booking.

import type { BookingResultEvent } from '@/booking/domain/events/bookingResultEvents.js';
import { CreditGiftCardPolicy } from '../../domain/policies/creditGiftCardPolicy.js';
import { GiftCardCrediting } from '../usecases/giftCardCrediting.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class BookingResultCrediting {
  private readonly _policy: CreditGiftCardPolicy;
  private readonly _useCase: GiftCardCrediting;

  constructor(policy: CreditGiftCardPolicy, useCase: GiftCardCrediting) {
    requireDependency(policy, "policy");
    requireDependency(useCase, "useCase");
    this._policy = policy;
    this._useCase = useCase;
  }

  handleBookingResults(event: BookingResultEvent): void {
    const command = this._policy.evaluate(event);
    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/services/BookingResultCrediting.java
package it.giannibombelli.wsc2026.giftcard.application.services;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardCrediting;
import it.giannibombelli.wsc2026.giftcard.application.policies.CreditGiftCardPolicy;

import static java.util.Objects.requireNonNull;


public class BookingResultCrediting {
    private final CreditGiftCardPolicy policy;
    private final GiftCardCrediting useCase;

    public BookingResultCrediting(CreditGiftCardPolicy creditGiftCardPolicy, GiftCardCrediting giftCardCrediting) {
        this.policy = requireNonNull(creditGiftCardPolicy);
        this.useCase = requireNonNull(giftCardCrediting);
    }

    public void handleBookingResults(BookingResultEvents events) {
        final CreditGiftCard cmd = policy.evaluate(events);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}
